### PR TITLE
{arch,plat}/x86: Refactor FS_BASE/GS_BASE operations

### PR DESCRIPTION
**Pull request by mogasergiu**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): `x86_64`
 - Platform(s): N/A
 - Application(s):N/A

### Description of changes

Depending on the configured CPU, `__FSGSBASE__` and its related built-ins may be defined. Use them when possible/available as it may improve performance when trying to read/write from such registers.

### CODE REVIEW

**Review by jschlumpp — APPROVED** [review]
Reviewed-by: Marco Schlumpp <marco@unikraft.io>

**Comment by mogasergiu** [comment]
Following an offline discussion with @skuenzer it was decided that we do not use intrinsics and instead we always use the newer instructions if available. Furthermore, all of this should be done with function pointers like ECTX. So at this point I am just going to call it a "refactor" and not bother with the details for simplicity's sake.

File: arch/x86/x86_64/fsgsbase.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+{
+	__asm__ __volatile__(
+		"wrfsbase	%0"
+		:
+		: "r" (fsbase)
+		: "memory"
+	);
+}
+
+
```
Comment (mogasergiu) [inline arch/x86/x86_64/fsgsbase.c:95]: Hmm interesting how checkpatch did not see this empty line. Will remove it on next relevant force push, I don't want to unnecessarily trigger another CI run.

File: arch/x86/Makefile.uk [hunk 2]
```diff
@@ -5,7 +5,13 @@ endif
 
 $(eval $(call addlib,libcontext))
 
+# TODO: Remove this dependency in the context of plat re-arch
```
Comment (michpappas) [inline arch/x86/Makefile.uk:8]: In the same spirit as in https://github.com/unikraft/unikraft/pull/1322#discussion_r1532162561, this code should not be in `arch`. Even more with the call to `UK_CTOR_PRIO`, as nothing in `arch` should depend on definitions in `lib`, `drivers`, or `include`. I would propose that we keep that code under `plat/` and later make it part of `libuklcpu`.

Also the commit message mentions `libcontext`, perhaps it needs to be updated.

Comment (michpappas) [reply arch/x86/Makefile.uk:8]: **Update:** During an offline discussion with @skuenzer it was agreed that we merge this now, as migration involves additional code that existed previously in `arch`, and do a migration once all depending PRs are merged.

**Review by michpappas — APPROVED** [review]
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

**Review by jschlumpp — APPROVED** [review]
Nice improvement.

Reviewed-by: Marco Schlumpp <marco@unikraft.io>

File: plat/common/include/x86/tls.h [hunk 3]
```diff
@@ -33,13 +33,8 @@
 #include <x86/cpu.h> /* rdmsrl, wrmsrl */
 #include <uk/arch/lcpu.h>
 
-/* TODO: On latest CPUs a faster instruction can be used:
- * https://www.kernel.org/doc/html/latest/x86/x86_64/fsgs.html
```
Comment (mogasergiu) [inline plat/common/include/x86/tls.h:37]: TODO: Add comment to give a heads up that this will be moved in the context of plat re-arch.

**Review by skuenzer — APPROVED** [review]
Nice work, looking forward to have this integrated.

Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request merged**